cpp/src/reductions/scan/ewm.cu

```diff
@@ -0,0 +1,374 @@
+/*
+ * Copyright (c) 2022-2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "scan.cuh"
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <cuda/functional>
+#include <thrust/scan.h>
+#include <thrust/transform_scan.h>
+
+namespace cudf {
+namespace detail {
+
+template <typename T>
+using pair_type = thrust::pair<T, T>;
+
+/**
+ * @brief functor to be summed over in a prefix sum such that
+ * the recurrence in question is solved. See
+ * G. E. Blelloch. Prefix sums and their applications. Technical Report
+ * CMU-CS-90-190, Nov. 1990. S. 1.4
+ * for details
+ */
+template <typename T>
+class recurrence_functor {
+ public:
+  __device__ pair_type<T> operator()(pair_type<T> ci, pair_type<T> cj)
+  {
+    return {ci.first * cj.first, ci.second * cj.first + cj.second};
+  }
+};
+
+template <typename T>
+class ewma_functor_base {
+ public:
+  T beta;
+  ewma_functor_base(T beta) : beta{beta} {}
+
+  pair_type<T> IDENTITY = {1.0, 0.0};
+};
+
+template <typename T, bool nulls, bool is_numerator>
+class ewma_adjust_functor : public ewma_functor_base<T> {
+  using ewma_functor_base<T>::ewma_functor_base;
+
+ public:
+  using tupletype = std::conditional_t<nulls, thrust::tuple<bool, int, T>, T>;
+
+  __device__ pair_type<T> operator()(tupletype const data)
+  {
+    if constexpr (nulls) {
```

Comment:
It looks like the nulls/non-nulls branches of this operator are completely distinct. Also, the `tupletype` above is conditioned on that. Can we just split these implementations into separate template specializations instead? `if constexpr` seems like an odd choice in this case since the functors are effectively completely different.

Reply:
Originally, I moved towards this approach of using `if constexpr` for the null case following your suggestion [in a previous round of review](https://github.com/rapidsai/cudf/pull/9027#discussion_r788193523). Since I've managed to miss precisely what you're requesting a couple of times, would you mind helping me out with a commit here @vyasr ? I'm happy to review.

Reply:
Yup, I remember. Sorry for the runaround. Things have changed a bit since then in part because we can use structured bindings now so a lot of the shared bits become trivial. Let me give this a shot.